File: Editor/Tools/CreateSceneTool.cs
using System;
using System.IO;
using System.Threading.Tasks;
using UnityEngine;
using UnityEditor;
using UnityEditor.SceneManagement;
using UnityEngine.SceneManagement;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for creating new scenes in Unity Editor
    /// </summary>
    public class CreateSceneTool : McpToolBase
    {
        public CreateSceneTool()
        {
            Name = "create_scene";
            Description = "Creates a new scene";
        }

        /// <summary>
        /// Creates a new scene with the provided parameters
        /// </summary>
        public override async Task<JObject> ExecuteAsync(JObject parameters)
        {
            try
            {
                // Validate parameters
                if (!parameters.TryGetValue("sceneName", out JToken sceneNameToken) || string.IsNullOrEmpty(sceneNameToken.ToString()))
                {
                    return JObject.FromObject(new
                    {
                        error = "sceneName is required"
                    });
                }

                string sceneName = sceneNameToken.ToString();
                string scenePath = parameters.TryGetValue("scenePath", out JToken scenePathToken) ? scenePathToken.ToString() : null;
                string templateScene = parameters.TryGetValue("templateScene", out JToken templateSceneToken) ? templateSceneToken.ToString() : null;

                // Check if we need to save current scene before creating a new one
                if (EditorSceneManager.GetActiveScene().isDirty)
                {
                    bool saveOk = await Task.Run(() => 
                    {
                        bool userAction = EditorUtility.DisplayDialog(
                            "Unsaved Changes",
                            "The current scene has unsaved changes. Save before creating a new scene?",
                            "Save",
                            "Don't Save"
                        );
                        
                        if (userAction)
                        {
                            return EditorSceneManager.SaveOpenScenes();
                        }
                        return true; // If user chooses not to save, we still proceed
                    });
                    
                    if (!saveOk)
                    {
                        return JObject.FromObject(new
                        {
                            error = "Failed to save current scene"
                        });
                    }
                }

                // Create a new scene
                Scene newScene;
                
                if (!string.IsNullOrEmpty(templateScene) && File.Exists(templateScene))
                {
                    // Create from template
                    bool loadResult = EditorSceneManager.OpenScene(templateScene, OpenSceneMode.Single) != null;
                    if (!loadResult)
                    {
                        return JObject.FromObject(new
                        {
                            error = $"Failed to load template scene: {templateScene}"
                        });
                    }
                    
                    newScene = EditorSceneManager.GetActiveScene();
                }
                else
                {
                    // Create empty scene
                    newScene = EditorSceneManager.NewScene(NewSceneSetup.DefaultGameObjects, NewSceneMode.Single);
                }
                
                // Save the scene if path is provided
                bool savedSuccessfully = true;
                string finalPath = scenePath;
                
                if (!string.IsNullOrEmpty(scenePath))
                {
                    // Ensure the directory exists
                    string directory = Path.GetDirectoryName(scenePath);
                    if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
                    {
                        Directory.CreateDirectory(directory);
                    }
                    
                    // Make sure path ends with .unity
                    if (!scenePath.EndsWith(".unity"))
                    {
                        finalPath = scenePath + ".unity";
                    }
                    
                    savedSuccessfully = EditorSceneManager.SaveScene(newScene, finalPath);
                }
                
                // Get scene info for response
                return JObject.FromObject(new
                {
                    success = true,
                    message = $"Scene '{sceneName}' created successfully",
                    saved = savedSuccessfully,
                    scene = new
                    {
                        name = newScene.name,
                        path = finalPath ?? "",
                        isLoaded = newScene.isLoaded,
                        isDirty = newScene.isDirty,
                        rootCount = newScene.rootCount
                    }
                });
            }
            catch (Exception ex)
            {
                Debug.LogError($"Error creating scene: {ex.Message}");
                return JObject.FromObject(new
                {
                    error = ex.Message
                });
            }
        }
    }
} 

File: Server/src/tools/createSceneTool.ts
import { z } from "zod";
import { Logger } from "../utils/logger.js";
import { McpUnity } from "../unity/mcpUnity.js";
import { ToolDefinition } from "./toolRegistry.js";
import { McpUnityError, ErrorType } from "../utils/errors.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

/**
 * Create a tool that creates a new scene in Unity
 */
export function createCreateSceneTool(
  mcpUnity: McpUnity,
  logger: Logger
): ToolDefinition {
  const toolName = "create_scene";
  return {
    name: toolName,
    description: "Creates a new scene in Unity",
    paramsSchema: z.object({
      sceneName: z.string().describe("Name for the new scene"),
      scenePath: z
        .string()
        .optional()
        .describe("Optional path where to save the scene"),
      templateScene: z
        .string()
        .optional()
        .describe("Optional template scene to use as a base"),
    }),
    handler: async (params: {
      sceneName: string;
      scenePath?: string;
      templateScene?: string;
    }): Promise<CallToolResult> => {
      logger.debug("Creating scene:", params);

      const response = await mcpUnity.sendRequest({
        method: toolName,
        params,
      });

      if (!response.success) {
        throw new McpUnityError(
          ErrorType.TOOL_EXECUTION,
          response.message || `Failed to create scene: ${params.sceneName}`
        );
      }

      return {
        success: true,
        message:
          response.message || `Scene created successfully: ${params.sceneName}`,
        content: [
          {
            type: "text",
            text:
              response.message ||
              `Scene ${params.sceneName} created successfully`,
          },
        ],
      };
    },
  };
}
